consensus/common/src/main/java/org/hyperledger/besu/consensus/common/bft/blockcreation/SchedulableMiningCoordinator.java

```diff
@@ -0,0 +1,139 @@
+
+  private static final Logger LOG = getLogger();
+
+  private final BftForksSchedule<MiningCoordinator> miningCoordinatorSchedule;
+  private final Blockchain blockchain;
+  private MiningCoordinator activeMiningCoordinator;
+  private long blockAddedObserverId;
+
+  public SchedulableMiningCoordinator(
+      final BftForksSchedule<MiningCoordinator> miningCoordinatorSchedule,
+      final Blockchain blockchain) {
+    this.miningCoordinatorSchedule = miningCoordinatorSchedule;
+    this.blockchain = blockchain;
+    this.activeMiningCoordinator =
+        this.miningCoordinatorSchedule.getFork(blockchain.getChainHeadBlockNumber()).getValue();
+  }
+
+  @Override
+  public void start() {
+    blockAddedObserverId = blockchain.observeBlockAdded(this);
+    activeMiningCoordinator.start();
+  }
+
+  @Override
+  public void stop() {
+    blockchain.removeObserver(blockAddedObserverId);
+    activeMiningCoordinator.stop();
+  }
+
+  @Override
+  public void awaitStop() throws InterruptedException {
+    activeMiningCoordinator.awaitStop();
+  }
+
+  @Override
+  public boolean enable() {
+    return activeMiningCoordinator.enable();
+  }
+
+  @Override
+  public boolean disable() {
+    return activeMiningCoordinator.disable();
+  }
+
+  @Override
+  public boolean isMining() {
+    return activeMiningCoordinator.isMining();
+  }
+
+  @Override
+  public Wei getMinTransactionGasPrice() {
+    return activeMiningCoordinator.getMinTransactionGasPrice();
+  }
+
+  @Override
+  public void setExtraData(final Bytes extraData) {
+    activeMiningCoordinator.setExtraData(extraData);
+  }
+
+  @Override
+  public Optional<Address> getCoinbase() {
+    return activeMiningCoordinator.getCoinbase();
+  }
+
+  @Override
+  public Optional<Block> createBlock(
+      final BlockHeader parentHeader,
+      final List<Transaction> transactions,
+      final List<BlockHeader> ommers) {
+    return activeMiningCoordinator.createBlock(parentHeader, transactions, ommers);
+  }
+
+  @Override
+  public void changeTargetGasLimit(final Long targetGasLimit) {
+    activeMiningCoordinator.changeTargetGasLimit(targetGasLimit);
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    final long currentBlock = event.getBlock().getHeader().getNumber();
+    final MiningCoordinator nextMiningCoordinator =
+        miningCoordinatorSchedule.getFork(currentBlock + 1).getValue();
+    if (activeMiningCoordinator != nextMiningCoordinator) {
+      LOG.debug(
+          "Switching mining coordinator at block {} from {} to {}",
+          currentBlock,
+          activeMiningCoordinator.getClass().getSimpleName(),
+          nextMiningCoordinator.getClass().getSimpleName());
+      activeMiningCoordinator.stop();
+      nextMiningCoordinator.start();
+      activeMiningCoordinator = nextMiningCoordinator;
+    }
+    if (activeMiningCoordinator instanceof BftMiningCoordinator) {
```

Comment:
rather than making this specific to BftMiningCoordinator I think we can can just conditionally cast a BlockAddedObserver then this become generic again.

Reply:
good point